### ISSUE TITLE
Kernel Module not loaded

### ISSUE THREAD

**Issue by nicedevil007** [linuxserver/docker-wireguard#47]
Hey there I'm running a Raspberry Pi 4 with DietPi installed on it.
That is based on debian and it is the 32bit version right now.

While trying to install your docker container via portainer I get this error:

![image](https://user-images.githubusercontent.com/17103076/88505487-e2e81400-cfd7-11ea-87fd-9d9fcb616c9b.png)

here is my stack content to deploy your image:

```
version: "3.7"

volumes:
  config:

services:
  app:
    image: linuxserver/wireguard:arm32v7-latest
    container_name: wireguard
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/Berlin
      - SERVERURL=testdomain.dedyn.io #optional
      - SERVERPORT=51820 #optional
      - PEERS=1 #optional
      - PEERDNS=auto #optional
      - INTERNAL_SUBNET=10.13.13.0 #optional
    volumes:
      - config:/config
      - /lib/modules:/lib/modules
    ports:
      - 51820:51820/udp
    sysctls:
      - net.ipv4.conf.all.src_valid_mark=1
    restart: unless-stopped
```

Am I missing something?

**Comment by nicedevil007** [comment]
ok... maybe I missed that I have to install the headers on the host?

`sudo apt install raspberrypi-kernel-headers`

**Comment by nicedevil007** [comment]
There error still persists :(

![image](https://user-images.githubusercontent.com/17103076/88530143-3589f600-d001-11ea-864a-c009ab0ffbd7.png)

**Comment by aptalca** [comment]
What is the kernel? `uname -a`

**Comment by nicedevil007** [comment]
`root@DietPi:~# uname -a
Linux DietPi 4.19.118-v7l+ #1311 SMP Mon Apr 27 14:26:42 BST 2020 armv7l GNU/Linux
`

**Comment by aptalca** [comment]
looks like the `raspberrypi-kernel-headers` package now only installs headers for the 5.4.51 kernel, but you're still on 4.19.118

Try a full apt update && upgrade then reboot and see if it updates your kernel

**Comment by nicedevil007** [comment]
thank you so far... to get the pi up to kernel 5 I need a rpi-update!

doing this right now. will report back!

**Comment by nicedevil007** [comment]
```
root@DietPi:~# uname -a
Linux DietPi 5.4.51-v7l+ #1328 SMP Mon Jul 27 16:57:05 BST 2020 armv7l GNU/Linux
```

Ok that worked for this issue... now I have a new one....

```
  LD [M]  /app/wireguard-linux-compat/src/wireguard.o
  Building modules, stage 2.
  MODPOST 1 modules
  CC [M]  /app/wireguard-linux-compat/src/wireguard.mod.o
  LD [M]  /app/wireguard-linux-compat/src/wireguard.ko
make: Leaving directory '/app/wireguard-linux-compat/src'
make: Entering directory '/app/wireguard-linux-compat/src'
  INSTALL /app/wireguard-linux-compat/src/wireguard.ko
  DEPMOD  5.4.51-v7l+
Warning: modules_install: missing 'System.map' file. Skipping depmod.
depmod -b "/" -a 5.4.51-v7l+
make: Leaving directory '/app/wireguard-linux-compat/src'
**** Let's test our new module. ****
RTNETLINK answers: Operation not permitted
**** The module is not active, review the logs. Sleeping now. . . ****
```

**Comment by nicedevil007** [comment]
@aptalca I saw that the images was updated and there was hope that this will fix my problem, unfurtunately not... Just want to let you know in case you tried to fix this issue?

**Comment by Patanouk** [comment]
Hi!
I had the exact same issue as you, trying to run the container on a RaspberryPi

I had the following volumes on my docker-compose file
```
volumes:
  - config:/config
  - ./lib-modules:/lib/modules
```

After deploying the stack / docker-compose up, the `lib-modules` folder didn't contain some of the needed files (modules.order && modules.builtin). 
It seems that the kernel headers installation done in the container is missing some of the files? Don't really have time to investigate further atm

The '/lib/modules/5.4.51-v7l+' folder (populated with 'sudo apt install raspberrypi-kernel-headers') did contain these files

Long story short, you can try to 
Remove your stack / docker-compose down
run `sudo apt reinstall raspberrypi-kernel-headers`
Double check that the previously missing files are now correctly there
Make sure that your container volume is linked to /lib/modules/
Restart your stack / docker-compose up

**Comment by tobbenb** [comment]
You didn't set the correct volume mount for /lib/modules. You did not map the host /lib/modules to the containers /lib/modules.

**Comment by Patanouk** [comment]
I know, and it does indeed work fine when I map the volume to the host one.

If you bind the container `/lib/modules` to an arbitrary empty folder (`./lib-volumes` in my case), the container will download the kernel headers at startup and populate the previously empty bind folder

However, it seems that the kernel headers which are downloaded that way are missing some of the files (see https://github.com/linuxserver/docker-wireguard/issues/47#issuecomment-665468964) for the detail of the missing files

Kernel headers downloaded via the package manager do contain the missing files

**Comment by nicedevil007** [comment]
@Patanouk thank you for your response, unfurtunately it didn't work for me. tryed your solution with an empty directory but that leads to a rejected container all the time...

```
root@DietPi:/lib/modules/5.4.51-v7l+# ls -la
total 2276
drwxr-xr-x  4 root root   4096 Jul 29 15:41 .
drwxr-xr-x 10 root root   4096 Jul 27 16:06 ..
lrwxrwxrwx  1 root root     34 Jul 23 16:31 build -> /usr/src/linux-headers-5.4.51-v7l+
drwxr-xr-x  2 root root   4096 Jul 27 19:21 extra
drwxr-xr-x 11 root root   4096 Jul 27 19:17 kernel
-rw-r--r--  1 root root 559450 Jul 29 15:41 modules.alias
-rw-r--r--  1 root root 580883 Jul 29 15:41 modules.alias.bin
-rw-r--r--  1 root root  12786 Jul 27 19:17 modules.builtin
-rw-r--r--  1 root root  14039 Jul 29 15:41 modules.builtin.bin
-rw-r--r--  1 root root  59057 Jul 27 19:17 modules.builtin.modinfo
-rw-r--r--  1 root root 191737 Jul 29 15:41 modules.dep
-rw-r--r--  1 root root 264099 Jul 29 15:41 modules.dep.bin
-rw-r--r--  1 root root    324 Jul 29 15:41 modules.devname
-rw-r--r--  1 root root  63033 Jul 27 19:17 modules.order
-rw-r--r--  1 root root    355 Jul 29 15:41 modules.softdep
-rw-r--r--  1 root root 241906 Jul 29 15:41 modules.symbols
-rw-r--r--  1 root root 296965 Jul 29 15:41 modules.symbols.bin
```

Isn't the `RTNET` thing a problem that should be fixed as well?

**Comment by JHenzi** [comment]
I'm not able to get them from the host and when the container tries to get them from the repos I see this;


**** No kernel headers found in the Ubuntu or Debian repos!! Will try the headers from host (if mapped), may or may not work ****,
**** Kernel headers don't seem to be available, can't compile the module. Sleeping now. . . ****

**Comment by j0nnymoe** [comment]
@JHenzi please post more information about your setup (host/hardware/kernel version etc) and what you've done to deploy the container. We can't work our your issue just from the error.

**Comment by aptalca** [comment]
If you're using a custom kernel, and there are no headers in ubuntu/debian repos, then you need to find the headers and install them on the host. Once you do that, you need to map `/usr/src:/usr/src`. The readme explains that.

**Comment by JHenzi** [comment]
Confused, why doesn't the container find any packages to install inside itself?

**Comment by aptalca** [comment]
Because you're using a custom kernel. The headers for that kernel are not published as a package in the ubuntu or debian repos
